// @wormhole-foundation/sdk-definitions/dist/esm/payloads/tokenBridge.js
import { range, } from "@wormhole-foundation/sdk-base";
import { payloadIdItem, chainItem, universalAddressItem, amountItem } from "../layout-items";
import { registerPayloadTypes } from "../vaa";
const fixedLengthStringItem = {
    binary: "bytes",
    size: 32,
    custom: {
        to: (val) => range(val.byteLength)
            .map((i) => String.fromCharCode(val[i]))
            .join(""),
        from: (str) => new Uint8Array(str.split("").map((c) => c.charCodeAt(0))),
    },
};
export const transferWithPayloadLayout = (customPayload) => [
    payloadIdItem(3),
    ...transferCommonLayout,
    { name: "from", ...universalAddressItem },
    { name: "payload", ...customPayload },
];
const transferCommonLayout = [
    {
        name: "token",
        binary: "object",
        layout: [
            { name: "amount", ...amountItem },
            { name: "address", ...universalAddressItem },
            { name: "chain", ...chainItem() },
        ],
    },
    {
        name: "to",
        binary: "object",
        layout: [
            { name: "address", ...universalAddressItem },
            { name: "chain", ...chainItem() },
        ],
    },
];
export const namedPayloads = [
    [
        "AttestMeta",
        [
            payloadIdItem(2),
            {
                name: "token",
                binary: "object",
                layout: [
                    { name: "address", ...universalAddressItem },
                    { name: "chain", ...chainItem() },
                ],
            },
            { name: "decimals", binary: "uint", size: 1 },
            { name: "symbol", ...fixedLengthStringItem },
            { name: "name", ...fixedLengthStringItem },
        ],
    ],
    ["Transfer", [payloadIdItem(1), ...transferCommonLayout, { name: "fee", ...amountItem }]],
    ["TransferWithPayload", transferWithPayloadLayout({ binary: "bytes" })],
];
registerPayloadTypes("TokenBridge", namedPayloads);
