// @wormhole-foundation/sdk-base/dist/cjs/utils/layout/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.layoutDiscriminator = exports.addFixedValues = exports.dynamicItemsOfLayout = exports.fixedItemsOfLayout = exports.deserializeLayout = exports.serializeLayout = exports.calcLayoutSize = void 0;
var size_1 = require("./size");
Object.defineProperty(exports, "calcLayoutSize", { enumerable: true, get: function () { return size_1.calcLayoutSize; } });
var serialize_1 = require("./serialize");
Object.defineProperty(exports, "serializeLayout", { enumerable: true, get: function () { return serialize_1.serializeLayout; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeLayout", { enumerable: true, get: function () { return deserialize_1.deserializeLayout; } });
var fixedDynamic_1 = require("./fixedDynamic");
Object.defineProperty(exports, "fixedItemsOfLayout", { enumerable: true, get: function () { return fixedDynamic_1.fixedItemsOfLayout; } });
Object.defineProperty(exports, "dynamicItemsOfLayout", { enumerable: true, get: function () { return fixedDynamic_1.dynamicItemsOfLayout; } });
Object.defineProperty(exports, "addFixedValues", { enumerable: true, get: function () { return fixedDynamic_1.addFixedValues; } });
var discriminate_1 = require("./discriminate");
Object.defineProperty(exports, "layoutDiscriminator", { enumerable: true, get: function () { return discriminate_1.layoutDiscriminator; } });
