// @wormhole-foundation/sdk-definitions/dist/cjs/payloads/automaticCircleBridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.namedPayloads = exports.connectPayload = exports.depositWithBytesPayload = exports.depositWithSizedLayoutPayload = void 0;
const layout_items_1 = require("../layout-items");
const vaa_1 = require("../vaa");
//from here: https://github.com/wormhole-foundation/wormhole-circle-integration/blob/105ad59bad687416527003e0241dee4020889341/evm/src/circle_integration/CircleIntegrationMessages.sol#L25
const depositWithPayloadBase = [
    (0, layout_items_1.payloadIdItem)(1),
    {
        name: "token",
        binary: "object",
        layout: [
            { name: "address", ...layout_items_1.universalAddressItem },
            { name: "amount", ...layout_items_1.amountItem },
        ],
    },
    { name: "sourceDomain", ...layout_items_1.circleDomainItem },
    { name: "targetDomain", ...layout_items_1.circleDomainItem },
    { name: "nonce", ...layout_items_1.circleNonceItem },
    { name: "caller", ...layout_items_1.universalAddressItem },
    { name: "mintRecipient", ...layout_items_1.universalAddressItem },
];
//a future optimization would be to calculate the layout size from the layout itself
//  thought that does require implementing arithmetic on number literals, which is its very
//  own can of worms
const depositWithSizedLayoutPayload = (byteSize, layout) => [
    ...depositWithPayloadBase,
    { name: "payloadSize", binary: "uint", size: 2, custom: byteSize, omit: true },
    { name: "payload", binary: "object", layout },
];
exports.depositWithSizedLayoutPayload = depositWithSizedLayoutPayload;
const depositWithBytesPayload = (customPayload) => [
    ...depositWithPayloadBase,
    { name: "payload", binary: "bytes", lengthSize: 2, ...customPayload },
];
exports.depositWithBytesPayload = depositWithBytesPayload;
//from here:
//  https://github.com/wormhole-foundation/example-circle-relayer/blob/189becd8d3935decb17383bd2e61b4909cbddc89/evm/src/circle-relayer/CircleRelayerMessages.sol#L16
exports.connectPayload = [
    (0, layout_items_1.payloadIdItem)(1),
    { name: "targetRelayerFee", ...layout_items_1.amountItem },
    { name: "toNativeTokenAmount", ...layout_items_1.amountItem },
    { name: "targetRecipient", ...layout_items_1.universalAddressItem },
];
exports.namedPayloads = [
    ["DepositWithPayload", (0, exports.depositWithBytesPayload)({})],
    ["TransferWithRelay", (0, exports.depositWithSizedLayoutPayload)(1 + 3 * 32, exports.connectPayload)],
];
(0, vaa_1.registerPayloadTypes)("AutomaticCircleBridge", exports.namedPayloads);
