// @wormhole-foundation/sdk-definitions/dist/cjs/payloads/bam.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.namedPayloads = exports.extendedMessageLayout = exports.tokenMessageLayout = exports.messageLayout = void 0;
const layout_items_1 = require("../layout-items");
const vaa_1 = require("../vaa");
const bamAddressItem = {
    binary: "bytes",
    lengthSize: 2,
};
const customOrEmpty = (custom) => (custom ? { custom } : {});
const messageLayout = (type, customContents) => [
    { name: "magicByte", binary: "uint", size: 1, custom: 0xbb, omit: true },
    { name: "version", binary: "uint", size: 1, custom: 0, omit: true },
    { name: "type", binary: "uint", size: 1, custom: type, omit: true },
    { name: "index", binary: "uint", size: 8 },
    { name: "targetChain", ...(0, layout_items_1.chainItem)() },
    { name: "targetAddress", ...bamAddressItem },
    { name: "senderAddress", ...bamAddressItem },
    { name: "contents", binary: "bytes", lengthSize: 2, ...customOrEmpty(customContents) },
];
exports.messageLayout = messageLayout;
const tokenMessageLayout = (custom) => [
    ...(0, exports.messageLayout)(1, custom?.contents),
    { name: "bridge", binary: "uint", size: 1, ...customOrEmpty(custom?.bridge) },
    { name: "assetIdentifier", binary: "uint", size: 2, ...customOrEmpty(custom?.assetIdentifier) },
    { name: "amount", ...layout_items_1.amountItem },
];
exports.tokenMessageLayout = tokenMessageLayout;
const extendedMessageLayout = (custom) => [
    ...(0, exports.messageLayout)(2, custom?.contents),
    { name: "relaySignal", binary: "bytes", lengthSize: 2, ...customOrEmpty(custom?.relaySignal) },
];
exports.extendedMessageLayout = extendedMessageLayout;
exports.namedPayloads = [
    ["Message", (0, exports.messageLayout)(0)],
    ["TokenMessage", (0, exports.tokenMessageLayout)()],
    ["ExtendedMessage", (0, exports.extendedMessageLayout)()],
];
(0, vaa_1.registerPayloadTypes)("BAM", exports.namedPayloads);
