// @wormhole-foundation/sdk-definitions/dist/cjs/payloads/tokenBridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.namedPayloads = exports.transferWithPayloadLayout = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
const layout_items_1 = require("../layout-items");
const vaa_1 = require("../vaa");
const fixedLengthStringItem = {
    binary: "bytes",
    size: 32,
    custom: {
        to: (val) => (0, sdk_base_1.range)(val.byteLength)
            .map((i) => String.fromCharCode(val[i]))
            .join(""),
        from: (str) => new Uint8Array(str.split("").map((c) => c.charCodeAt(0))),
    },
};
const transferWithPayloadLayout = (customPayload) => [
    (0, layout_items_1.payloadIdItem)(3),
    ...transferCommonLayout,
    { name: "from", ...layout_items_1.universalAddressItem },
    { name: "payload", ...customPayload },
];
exports.transferWithPayloadLayout = transferWithPayloadLayout;
const transferCommonLayout = [
    {
        name: "token",
        binary: "object",
        layout: [
            { name: "amount", ...layout_items_1.amountItem },
            { name: "address", ...layout_items_1.universalAddressItem },
            { name: "chain", ...(0, layout_items_1.chainItem)() },
        ],
    },
    {
        name: "to",
        binary: "object",
        layout: [
            { name: "address", ...layout_items_1.universalAddressItem },
            { name: "chain", ...(0, layout_items_1.chainItem)() },
        ],
    },
];
exports.namedPayloads = [
    [
        "AttestMeta",
        [
            (0, layout_items_1.payloadIdItem)(2),
            {
                name: "token",
                binary: "object",
                layout: [
                    { name: "address", ...layout_items_1.universalAddressItem },
                    { name: "chain", ...(0, layout_items_1.chainItem)() },
                ],
            },
            { name: "decimals", binary: "uint", size: 1 },
            { name: "symbol", ...fixedLengthStringItem },
            { name: "name", ...fixedLengthStringItem },
        ],
    ],
    ["Transfer", [(0, layout_items_1.payloadIdItem)(1), ...transferCommonLayout, { name: "fee", ...layout_items_1.amountItem }]],
    ["TransferWithPayload", (0, exports.transferWithPayloadLayout)({ binary: "bytes" })],
];
(0, vaa_1.registerPayloadTypes)("TokenBridge", exports.namedPayloads);
