// @wormhole-foundation/connect-sdk/dist/cjs/routes/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteTransferRequest = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
const sdk_definitions_1 = require("@wormhole-foundation/sdk-definitions");
class RouteTransferRequest {
    from;
    to;
    source;
    destination;
    fromChain;
    toChain;
    constructor(from, to, fromChain, toChain, source, destination) {
        this.from = from;
        this.fromChain = fromChain;
        this.to = to;
        this.toChain = toChain;
        this.source = source;
        this.destination = destination;
    }
    normalizeAmount(amount) {
        return (0, sdk_base_1.normalizeAmount)(amount, BigInt(this.source.decimals));
    }
    displayAmount(amount) {
        return (0, sdk_base_1.displayAmount)(amount, BigInt(this.source.decimals), BigInt(this.source.decimals));
    }
    static async create(wh, params, fromChain, toChain) {
        fromChain = fromChain ?? wh.getChain(params.from.chain);
        toChain = toChain ?? wh.getChain(params.to.chain);
        const sourceDetails = await getTokenDetails(fromChain, params.source);
        const destDetails = params.destination
            ? await getTokenDetails(toChain, params.destination)
            : undefined;
        const rtr = new RouteTransferRequest(params.from, params.to, fromChain, toChain, sourceDetails, destDetails);
        return rtr;
    }
}
exports.RouteTransferRequest = RouteTransferRequest;
async function getTokenDetails(chain, token) {
    const address = (0, sdk_definitions_1.isTokenId)(token) ? (0, sdk_definitions_1.canonicalAddress)(token) : token;
    const details = chain.config.tokenMap
        ? sdk_base_1.tokens.filters.byAddress(chain.config.tokenMap, address)
        : undefined;
    const symbol = details ? details.symbol : undefined;
    const wrapped = token === "native" ? await chain.getNativeWrappedTokenId() : undefined;
    const decimals = Number(await chain.getDecimals((0, sdk_definitions_1.isTokenId)(token) ? token.address : token));
    return {
        id: token,
        decimals,
        wrapped,
        symbol,
    };
}
